Build(Test) error on version 2.3.0 with JDK 21

**SpringDocApp193Test.java** fails with **JDK 21** 

Version:
``` 
	<artifactId>springdoc-openapi</artifactId>
	<version>2.3.0</version>
```

`mvn clean package `  fails with:
```
java.lang.AssertionError: components.schemas.Books.allOf[1].properties
Unexpected: first
 ; components.schemas.Books.allOf[1].properties
Unexpected: last
 ; components.schemas
Unexpected: Book
```

Screenshot:

![image](https://github.com/springdoc/springdoc-openapi/assets/3501456/e47614f4-7187-4f33-a8c3-5af2f50fa22b)

_On JDK 17 build passes without a problem._

# Minimal Reproducible Example:

1. Clone the repository:

```bash
git clone https://github.com/springdoc/springdoc-openapi.git
```

3. Install [SDKMAN](https://sdkman.io/install).
4. Install Maven, Java 17 and Java 21:

```bash
sdk install maven &&
sdk install java 17.0.10-tem &&
sdk install java 21.0.2-tem
```

5. Set Java 17 as the default Java version:

```bash
sdk default java 17.0.10-tem
```

6. Run that specific test:

```bash
mvn test -Dtest=SpringDocApp193Test "-Dsurefire.failIfNoSpecifiedTests=false"
# Passes
```

7. Switch to Java 21:

```bash
sdk default java 21.0.2-tem
```

8. Run that specific test:

```java
mvn test -Dtest=SpringDocApp193Test "-Dsurefire.failIfNoSpecifiedTests=false"
# Fails

[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running test.org.springdoc.api.v30.app193.SpringDocApp193Test
WARNING: A Java agent has been loaded dynamically (/home/user/.m2/repository/net/bytebuddy/byte-buddy-agent/1.14.12/byte-buddy-agent-1.14.12.jar)
WARNING: If a serviceability tool is in use, please run with -XX:+EnableDynamicAgentLoading to hide this warning
WARNING: If a serviceability tool is not in use, please run with -Djdk.instrument.traceUsage for more information
WARNING: Dynamic loading of agents will be disallowed by default in a future release
OpenJDK 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended

MockHttpServletRequest:
      HTTP Method = GET
      Request URI = /v3/api-docs
       Parameters = {}
          Headers = []
             Body = null
    Session Attrs = {}

Handler:
             Type = org.springdoc.webmvc.api.OpenApiWebMvcResource
           Method = org.springdoc.webmvc.api.OpenApiWebMvcResource#openapiJson(HttpServletRequest, String, Locale)

Async:
    Async started = false
     Async result = null

Resolved Exception:
             Type = null

ModelAndView:
        View name = null
             View = null
            Model = null

FlashMap:
       Attributes = null

MockHttpServletResponse:
           Status = 200
    Error message = null
          Headers = [Content-Type:"application/json", Content-Length:"1621"]
     Content type = application/json
             Body = {"openapi":"3.0.1","info":{"title":"OpenAPI definition","version":"v0"},"servers":[{"url":"http://localhost","description":"Generated server url"}],"paths":{"/test":{"get":{"tags":["basic-controller"],"summary":"get","description":"Provides a list of books.","operationId":"get","responses":{"200":{"description":"OK","content":{"*/*":{"schema":{"oneOf":[{"$ref":"#/components/schemas/Books"}]}}}}}}},"/test1":{"get":{"tags":["basic-controller"],"summary":"get1","description":"Provides an animal.","operationId":"get1","responses":{"200":{"description":"OK","content":{"*/*":{"schema":{"oneOf":[{"$ref":"#/components/schemas/Cat"},{"$ref":"#/components/schemas/Dog"}]}}}}}}}},"components":{"schemas":{"Book":{"type":"object","properties":{"title":{"type":"string"}},"description":"Represents a Book."},"Books":{"type":"array","description":"Represents a list of Books.","allOf":[{"$ref":"#/components/schemas/Knowledge"},{"type":"object","properties":{"empty":{"type":"boolean"},"first":{"$ref":"#/components/schemas/Book"},"last":{"$ref":"#/components/schemas/Book"}}}]},"Knowledge":{"type":"object","description":"Represents the knowledge."},"Animal":{"type":"object","description":"Represents an Animal class."},"Cat":{"type":"object","description":"Represents a Cat class.","allOf":[{"$ref":"#/components/schemas/Animal"},{"type":"object","properties":{"speed":{"type":"integer","format":"int32"}}}]},"Dog":{"type":"object","description":"Represents a Dog class.","allOf":[{"$ref":"#/components/schemas/Animal"},{"type":"object","properties":{"name":{"type":"string"},"age":{"type":"integer","format":"int32"}}}]}}}}
    Forwarded URL = null
   Redirected URL = null
          Cookies = []
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.778 s <<< FAILURE! -- in test.org.springdoc.api.v30.app193.SpringDocApp193Test
[ERROR] test.org.springdoc.api.v30.app193.SpringDocApp193Test.testApp -- Time elapsed: 0.893 s <<< FAILURE!
java.lang.AssertionError: 
components.schemas.Books.allOf[1].properties
Unexpected: first
 ; components.schemas.Books.allOf[1].properties
Unexpected: last
 ; components.schemas
Unexpected: Book

        at org.skyscreamer.jsonassert.JSONAssert.assertEquals(JSONAssert.java:417)
        at org.skyscreamer.jsonassert.JSONAssert.assertEquals(JSONAssert.java:394)
        at org.skyscreamer.jsonassert.JSONAssert.assertEquals(JSONAssert.java:336)
        at test.org.springdoc.api.v30.AbstractSpringDocV30Test.testApp(AbstractSpringDocV30Test.java:53)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   SpringDocApp193Test>AbstractSpringDocV30Test.testApp:53 components.schemas.Books.allOf[1].properties
Unexpected: first
 ; components.schemas.Books.allOf[1].properties
Unexpected: last
 ; components.schemas
Unexpected: Book

[INFO] 
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Spring openapi documentation 2.4.1-SNAPSHOT:
[INFO] 
[INFO] Spring openapi documentation ....................... SUCCESS [  0.129 s]
[INFO] springdoc-openapi-starter-common ................... SUCCESS [  5.227 s]
[INFO] springdoc-openapi-starter-webmvc-api ............... FAILURE [  5.492 s]
[INFO] springdoc-openapi-starter-webflux-api .............. SKIPPED
[INFO] springdoc-openapi-starter-webmvc-ui ................ SKIPPED
[INFO] springdoc-openapi-starter-webflux-ui ............... SKIPPED
[INFO] springdoc-openapi-tests ............................ SKIPPED
[INFO] springdoc-openapi-security-tests ................... SKIPPED
[INFO] springdoc-openapi-groovy-tests ..................... SKIPPED
[INFO] springdoc-openapi-javadoc-tests .................... SKIPPED
[INFO] springdoc-openapi-function-webmvc-tests ............ SKIPPED
[INFO] springdoc-openapi-function-webflux-tests ........... SKIPPED
[INFO] springdoc-openapi-actuator-webflux-tests ........... SKIPPED
[INFO] springdoc-openapi-actuator-webmvc-tests ............ SKIPPED
[INFO] springdoc-openapi-kotlin-webflux-tests ............. SKIPPED
[INFO] springdoc-openapi-kotlin-webmvc-tests .............. SKIPPED
[INFO] springdoc-openapi-hateoas-tests .................... SKIPPED
[INFO] springdoc-openapi-data-rest-tests .................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  11.136 s
[INFO] Finished at: 2024-03-19T13:37:53-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.1.2:test (default-test) on project springdoc-openapi-starter-webmvc-api: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/user/projects/springdoc-openapi/springdoc-openapi-starter-webmvc-api/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :springdoc-openapi-starter-webmvc-api
```

--- a/springdoc-openapi-starter-webmvc-api/src/test/java/test/org/springdoc/api/v30/app193/SpringDocApp193Test.java
+++ b/springdoc-openapi-starter-webmvc-api/src/test/java/test/org/springdoc/api/v30/app193/SpringDocApp193Test.java
@@ -23,12 +23,37 @@
 package test.org.springdoc.api.v30.app193;
 
 
-import test.org.springdoc.api.v30.AbstractSpringDocV30Test;
+import org.apache.commons.lang3.JavaVersion;
+import org.junit.jupiter.api.Test;
+import org.springdoc.core.utils.Constants;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.web.servlet.MvcResult;
+import test.org.springdoc.api.AbstractCommonTest;
 
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 
+import static org.apache.commons.lang3.SystemUtils.isJavaVersionAtLeast;
+import static org.hamcrest.Matchers.is;
+import static org.skyscreamer.jsonassert.JSONAssert.assertEquals;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 
-public class SpringDocApp193Test extends AbstractSpringDocV30Test {
+@SpringBootTest
+public class SpringDocApp193Test extends AbstractCommonTest {
+
+	@Test
+	public void testApp() throws Exception {
+		final MvcResult mockMvcResult = mockMvc.perform(get(Constants.DEFAULT_API_DOCS_URL)).andExpect(status().isOk())
+				.andExpect(jsonPath("$.openapi", is("3.0.1"))).andReturn();
+		final String result = mockMvcResult.getResponse().getContentAsString();
+		// In Java 21 the getFirst() and getLast() methods were added to the List interface.
+		// Those are the POJO getters, therefore Jackson will add them during serialization.
+		// So there are two different expected results for Java prior 21 and starting from Java 21.
+		final var expectedResponseFile = isJavaVersionAtLeast(JavaVersion.JAVA_21) ? "app193-1.json" : "app193.json";
+		final String expected = getContent("results/3.0.1/" + expectedResponseFile);
+		assertEquals(expected, result, true);
+	}
 
 	@SpringBootApplication
 	static class SpringDocTestApp {}

--- a/springdoc-openapi-starter-webmvc-api/src/test/resources/results/3.0.1/app193-1.json
+++ b/springdoc-openapi-starter-webmvc-api/src/test/resources/results/3.0.1/app193-1.json
@@ -1,0 +1,158 @@
+{
+  "openapi": "3.0.1",
+  "info": {
+    "title": "OpenAPI definition",
+    "version": "v0"
+  },
+  "servers": [
+    {
+      "url": "http://localhost",
+      "description": "Generated server url"
+    }
+  ],
+  "paths": {
+    "/test": {
+      "get": {
+        "tags": [
+          "basic-controller"
+        ],
+        "summary": "get",
+        "description": "Provides a list of books.",
+        "operationId": "get",
+        "responses": {
+          "200": {
+            "description": "OK",
+            "content": {
+              "*/*": {
+                "schema": {
+                  "oneOf": [
+                    {
+                      "$ref": "#/components/schemas/Books"
+                    }
+                  ]
+                }
+              }
+            }
+          }
+        }
+      }
+    },
+    "/test1": {
+      "get": {
+        "tags": [
+          "basic-controller"
+        ],
+        "summary": "get1",
+        "description": "Provides an animal.",
+        "operationId": "get1",
+        "responses": {
+          "200": {
+            "description": "OK",
+            "content": {
+              "*/*": {
+                "schema": {
+                  "oneOf": [
+                    {
+                      "$ref": "#/components/schemas/Cat"
+                    },
+                    {
+                      "$ref": "#/components/schemas/Dog"
+                    }
+                  ]
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  },
+  "components": {
+    "schemas": {
+      "Book":
+      {
+        "type": "object",
+        "properties":
+        {
+          "title":
+          {
+            "type": "string"
+          }
+        },
+        "description": "Represents a Book."
+      },
+      "Books": {
+        "type": "array",
+        "description": "Represents a list of Books.",
+        "allOf": [
+          {
+            "$ref": "#/components/schemas/Knowledge"
+          },
+          {
+            "type": "object",
+            "properties": {
+              "empty": {
+                "type": "boolean"
+              },
+              "first":
+              {
+                "$ref": "#/components/schemas/Book"
+              },
+              "last":
+              {
+                "$ref": "#/components/schemas/Book"
+              }
+            }
+          }
+        ]
+      },
+      "Knowledge": {
+        "type": "object",
+        "description": "Represents the knowledge."
+      },
+      "Animal": {
+        "type": "object",
+        "description": "Represents an Animal class."
+      },
+      "Cat": {
+        "type": "object",
+        "description": "Represents a Cat class.",
+        "allOf": [
+          {
+            "$ref": "#/components/schemas/Animal"
+          },
+          {
+            "type": "object",
+            "properties": {
+              "speed": {
+                "type": "integer",
+                "format": "int32"
+              }
+            }
+          }
+        ]
+      },
+      "Dog": {
+        "type": "object",
+        "description": "Represents a Dog class.",
+        "allOf": [
+          {
+            "$ref": "#/components/schemas/Animal"
+          },
+          {
+            "type": "object",
+            "properties": {
+              "name": {
+                "type": "string"
+              },
+              "age": {
+                "type": "integer",
+                "format": "int32"
+              }
+            }
+          }
+        ]
+      }
+    }
+  }
+}

